### ISSUE TITLE
[Hold for payment 2024-03-06] [$500] Workspace-The archived task report is no longer archived after reopening the report.

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#29049]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Launch app
2. Tap announce room
3. Tap plus icon
4. Select assign task
5. Enter title "Ree task" and tap next
6. Tap confirm task
7. Navigate back to LHN
8. Tap profile
9. Tap Workspaces
10. Select the Workspace(in whose announce room task is created)
11. Tap on 3dots on top and delete the workspace
12. Navigate to LHN and note the created task( " Ree Task") is archived
13. Tap on archived Workspace and confirm Workspace is archived
14. Open the archived announce room
15. Tap on the task and note the task report header
16. Tap back and tick the task
17. Tap on the task again and note the task report header

## Expected Result:
In archived WS-announce room, user must not be able to tick the task and interact.
The archived task report must remain archived after reopening the report also.

## Actual Result:
In archived WS-announce room, user able to tick the task and interact.
The archived task report is no longer archived after reopening the report.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.79-3

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/632d9337-1085-4a50-b2c5-ef81d699d57f

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause-Internal Team

**Slack conversation:** @

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019a5d1f4f0cd37991</li>
        <li>Upwork Job ID: 1710593181054087168</li>
        <li>Last Price Increase: 2023-11-20</li>
<li>Automatic offers: </li>
<ul>
<li>mollfpr | Reviewer | 27773918</li>
<li>s-alves10 | Contributor | 27773920</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @muttmuure</details>

**Comment by graylewis** [comment]
###Proposal
###Please re-state the problem that we are trying to solve in this issue.
Tasks that belong to an archived task report are still able to be completed/reopened

###What is the root cause of that problem?
This actually seems to be a backend issue. This is evident due to the data coming from the API from the openReport call containing the property: `"isDeletedParentAction": false`. The reason the title on the task is briefly correct [shows {TaskName} (Archived)] is due to optimistic updates. The task report is never actually archived. 

###What changes do you think we should make in order to solve the problem?
Backend code needs to be checked for bugs regarding the calculation of the isDeletedParentAction prop.

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The task report action in archived announce room is interactable and the task report is no longer archived after reopen the report

### What is the root cause of that problem?

When we delete a workspace, we set the archived state of related reports here
https://github.com/Expensify/App/blob/389d7b0c4b96c6f2a2295055c685791d35b65252/src/libs/actions/Policy.js#L93-L102
Task report is marked as archived as well. But when we reopen the task report, the backend returns `state: 0, status: 0` again and so the task report returns to the normal state. This is one reason

After the workspace is deleted, we should disable the interaction of tasks. We disable the task using the below function
https://github.com/Expensify/App/blob/389d7b0c4b96c6f2a2295055c685791d35b65252/src/libs/actions/Task.js#L889-L903
As you can see, we don't check if the parent report is an archived room. This is another reason

### What changes do you think we should make in order to solve the problem?

This issue requires both the backend and frontend fixes
1. Backend
When a workspace is deleted, the backend should mark all the reports related to the workspace `archived`

2. Frontend
We need to update the `canModifyTask` function above to check the parent report
Update the above function as follows
```
function canModifyTask(taskReport, sessionAccountID) {
    if (ReportUtils.isCanceledTaskReport(taskReport)) {
        return false;
    }

    const parentReport = ReportUtils.getParentReport(taskReport);
    if (ReportUtils.isArchivedRoom(parentReport)) {
        return false;
    }

    if (sessionAccountID === getTaskOwnerAccountID(taskReport) || sessionAccountID === getTaskAssigneeAccountID(taskReport)) {
        return true;
    }

    // If you don't have access to the task report (maybe haven't opened it yet), check if you can access the parent report
    // - If the parent report is an #admins only room
    // - If you are a policy admin
    return ReportUtils.isAllowedToComment(parentReport);
}
```

This works as expected

### What alternative solutions did you explore? (Optional)

**Comment by mollfpr** [comment]
Asking the internal team for the expected behavior of the task report when the parent report is archived.

**Comment by muttmuure** [comment]
Did you find anything out @mollfpr ?

**Comment by mollfpr** [comment]
Not overdue, still no update on the thread, I'll bump it once again.

**Comment by mollfpr** [comment]
Bumping the Slack again to get some help.

@s-alves10 For the `canModifyTask` logic, I think it's better to check the status instead of the parent report archived.

**Comment by s-alves10** [comment]
@mollfpr 

I think we need to confirm the expected behavior. 

> For the canModifyTask logic, I think it's better to check the status instead of the parent report archived.

Yes, you're right. But the status is changed after opening the report because backend returns `state: 0, status: 0`. Also the expected behavior is _In archived WS-announce room, user must not be able to tick the task and interact._

**Comment by mollfpr** [comment]
I've got feedback, and the task report should behave as the nested thread works. So, even if the parent report is archived, the task report is still `OPEN`.

> Yes, you're right. But the status is changed after opening the report because backend returns state: 0, status: 0.

The thread report status should not be changed on deleting the workspace.

---

>In archived WS-announce room, user must not be able to tick the task and interact.

@muttmuure Could you confirm if the above expectation is still needed? For me, it does not make sense, because the task report is still `OPEN` and the user can't interact with it.

>The archived task report must remain archived after reopening the report also.

We can remove this expected result.

The only thing we can do in this issue is, fix the LHN showing that the task report is archived after deleting the workspace.

@muttmuure Let me know what you think from the above. Thank you!

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The task reports linked to a workspace  appear as archived when we delete the workspace

### What is the root cause of that problem?

When we delete a workspace we close all reports linked to this workspace [here ](https://github.com/Expensify/App/blob/e99c297332c738475cfe5701e0a00ba0b5d58029/src/libs/actions/Policy.js#L133-L149)

We never filter those reports to not include tasks.


### What changes do you think we should make in order to solve the problem?

We should filter the reports that we are going to close to not include the task reports.

For example we can do that here:
https://github.com/Expensify/App/blob/e99c297332c738475cfe5701e0a00ba0b5d58029/src/pages/workspace/WorkspaceInitialPage.js#L82
by adding the condition: `&& report.type != CONST.REPORT.TYPE.TASK`

It works well.


### What alternative solutions did you explore? (Optional)

Otherwise we can do it directly in the method deleteWorkspace with the same  condition.

**Comment by mollfpr** [comment]
Friendly bump @muttmuure for https://github.com/Expensify/App/issues/29049#issuecomment-1767496708

**Comment by s-alves10** [comment]
@muttmuure @mollfpr 

We're now archiving all the reports related to the workspace when deleting a workspace. Now expectation has changed, but I don't think this isn't an issue related only to task. I think we need to set expectation clearly: What to archive and what not to archive.

**Comment by MariaHCD** [comment]
Coming from: https://github.com/Expensify/App/issues/28187

We should definitely iron out what reports should be optimistically archived or not when a workspace is deleted. Not super sure about task reports but from this [slack conversation](https://expensify.slack.com/archives/C01GTK53T8Q/p1698754831118739), expense reports should not be archived if the workspace is deleted. Users can still continue to comment on their financial transactions.

**Comment by muttmuure** [comment]
I will defer to @MariaHCD's opinion here. Reports should not be archived when a workspace is deleted

**Comment by muttmuure** [comment]
working on higher priorities

**Comment by mollfpr** [comment]
Asking clarification on the Slack [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1699533469444789?thread_ts=1698754831.118739&cid=C01GTK53T8Q).

**Comment by mollfpr** [comment]
Sorry for the delay 🙏

Let's go with @s-alves10 [proposal](https://github.com/Expensify/App/issues/29049#issuecomment-1751714379), but we must also change the API response.

🎀 👀 🎀 C+ reviewed!

**Comment by pecanoro** [comment]
It seems this is a problem with all types of child reports since money requests get stuck in a weird limbo as well:

![image](https://github.com/Expensify/App/assets/6474442/9d0d9081-0bdc-48ba-97e0-9ffc4c643c4e)

**Comment by pecanoro** [comment]
It seems this is a bigger problem with deleting workspaces, so I brought it up internally to see how we should handle in the back-end

**Comment by s-alves10** [comment]
@mollfpr 

The backend should be fixed in order to fix this issue completely. The task report would remain archived only after the backend is fixed.

As a workaround, we can update the [ReportUtils.isArchivedRoom](https://github.com/Expensify/App/blob/85f372c4e35c30119e72c7689959282a298bad1b/src/libs/ReportUtils.js#L640-L642) function to check the parent report as well, but I don't think this is a solution.

Will you let me know when I can expect the backend fix?

**Comment by mollfpr** [comment]
We should wait for the BE update, and I think @pecanoro will let us know when it's done.

@muttmuure Let's put this issue on hold.

**Comment by neil-marcellini** [comment]
Hey @pecanoro it looks like you haven't gotten started so I'm going to work on a solution to my issue which is very much related. I'll archive all reports when a workspace is deleted, which should fix this too.

**Comment by pecanoro** [comment]
Not overdue, @neil-marcellini is working on a PR to fix this.

**Comment by pecanoro** [comment]
Not overdue, @neil-marcellini is working on a PR to fix this.

**Comment by neil-marcellini** [comment]
Yep, still the same story. The PR is pretty close to ready for review.

**Comment by neil-marcellini** [comment]
❗ 🔥 Heads up, I'm going to be OOO working 0% from 1/1-1/12/24. I'll be back on 1/15/24. I'm planning to work 50% tomorrow 12/29. Please feel free to un-assign, re-assign, whatever. When I return I'll pick back up whatever I can and do my best to GSD in order of priority. I will also post all of my assigned issues in #engineering-chat in Slack to try to recruit volunteers.

**Comment by pecanoro** [comment]
@neil-marcellini is OOO so no update this week.

**Comment by neil-marcellini** [comment]
The PR went through the first round of reviews. I'm going to do a bit more manually testing and make sure the proper updates are pushed to users, then put it up for review again soon.

**Comment by neil-marcellini** [comment]
This should be fixed now. @muttmuure would you please verify if it's fixed and let us know if anything else is needed?

**Comment by s-alves10** [comment]
@mollfpr @muttmuure 
PR is ready for review: https://github.com/Expensify/App/pull/31674

**Comment by s-alves10** [comment]
@pecanoro 

Automation looks not working. Pay day would be 6th, Mar here

**Comment by muttmuure** [comment]
Seems like this is fixed, issuing payment now

**Comment by muttmuure** [comment]
Sent new offer to @mollfpr since the original offer expired

**Comment by muttmuure** [comment]
@s-alves10 has been paid

**Comment by mollfpr** [comment]
> [@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
> [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

No offending PR was found. This is an improvement and need a backend change to fix the issue both ways.

> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

The regression step should be good.

> [@mollfpr] Determine if we should create a regression test for this bug.
 [@mollfpr] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

1. Open ND and log in with any account
2. Go to announce room
3. Create a task
4. Delete the workspace in whose announce room the task is just created
5. Note that the task report was archived in LHN
6. Tap task in the archived announce room
7. Verify that the task report remains archived and a user isn't able to tick the task to interact
8. 👍 or 👎 

---

@muttmuure I'll request in NewDot. Could you create the payment summary? Thank you!

**Comment by muttmuure** [comment]
$500 - @mollfpr C+

**Comment by JmillsExpensify** [comment]
$500 approved for @mollfpr based on summary.
